auto-value-gson/src/main/java/com/ryanharter/auto/value/gson/AutoValueGsonExtension.java

```diff
@@ -671,7 +708,166 @@ private MethodSpec createReadMethod(ClassName className,
     return readMethod.build();
   }
 
-  /** Returns a default value for initializing well-known types, or else {@code null}. */
+  private MethodSpec createReadMethod(ClassName className,
+      TypeName autoValueClassName,
+      List<Property> properties,
+      ImmutableMap<TypeName, FieldSpec> adapters,
+      ClassName jsonAdapter,
+      List<TypeVariableName> typeParams,
+      BuilderContext builderContext,
+      ProcessingEnvironment processingEnvironment) {
+    ParameterSpec jsonReader = ParameterSpec.builder(JsonReader.class, "jsonReader").build();
+    MethodSpec.Builder readMethod = MethodSpec.methodBuilder("read")
+        .addAnnotation(Override.class)
+        .addAnnotation(AnnotationSpec.builder(SuppressWarnings.class)
+            .addMember("value", "\"unchecked\"")
+            .build())
+        .addModifiers(PUBLIC)
+        .returns(autoValueClassName)
+        .addParameter(jsonReader)
+        .addException(IOException.class);
+
+    ClassName token = ClassName.get(JsonToken.NULL.getDeclaringClass());
+
+    readMethod.beginControlFlow("if ($N.peek() == $T.NULL)", jsonReader, token);
+    readMethod.addStatement("$N.nextNull()", jsonReader);
+    readMethod.addStatement("return null");
+    readMethod.endControlFlow();
+
+
+    // Instantiate the AutoValue'd class builder
+    FieldSpec builder = FieldSpec.builder(ClassName.get(builderContext.builderType()), "builder")
+        .build();
+
+    Set<ExecutableElement> builderMethods = builderContext.builderMethods();
```

Comment:
How do we handle parameters? A common pattern for builder factory methods is to accept required properties as its parameters.

Reply:
`BuilderContext#builderMethods` only returns static no-arg factory methods, so shouldn't be an issue. But perhaps worth calling out in the README?

Reply:
Gotcha, yeah a readme bit would be good